rosbag2_compression/src/rosbag2_compression/zstd_decompressor.cpp

```diff
@@ -0,0 +1,242 @@
+
+  // Read in buffer, handling accordingly
+  const auto file_pointer = open_file(uri.c_str(), "rb");
+  if (file_pointer == nullptr) {
+    std::stringstream errmsg;
+    errmsg << "Failed to open file: \"" << uri << "\" for binary reading!";
+
+    throw std::runtime_error{errmsg.str()};
+  }
+
+  // Initialize compress_buffer with size = compressed_buffer_length.
+  // Uniform initialization cannot be used here since it will choose
+  // the initializer list constructor instead.
+  std::vector<uint8_t> compressed_buffer(compressed_buffer_length);
+
+  const auto nRead = fread(
+    compressed_buffer.data(), sizeof(decltype(compressed_buffer)::value_type),
+    compressed_buffer_length, file_pointer);
+
+  if (nRead != compressed_buffer_length) {
+    ROSBAG2_COMPRESSION_LOG_ERROR_STREAM("Bytes read !(" <<
+      nRead << ") != compressed_buffer_length (" << compressed_buffer_length <<
+      ")!");
+    // An error indicator is set by fread, so the following check will throw.
+  }
+
+  if (ferror(file_pointer)) {
+    fclose(file_pointer);
+
+    std::stringstream errmsg;
+    errmsg << "Unable to read binary data from file: \"" << uri << "\"!";
+
+    throw std::runtime_error{errmsg.str()};
+  }
+
+  fclose(file_pointer);
+  return compressed_buffer;
+}
+
+/**
+ * Writes the output buffer to the specified file path.
+ * \param output_buffer is the data to write.
+ * \param uri is the relative file path to the output storage.
+ */
+void write_output_buffer(
+  const std::vector<uint8_t> & output_buffer,
+  const std::string & uri)
+{
+  const auto file_pointer = open_file(uri.c_str(), "wb");
+  if (file_pointer == nullptr) {
+    std::stringstream errmsg;
+    errmsg << "Failed to open file: \"" << uri << "\" for binary writing!";
+
+    throw std::runtime_error{errmsg.str()};
+  }
+
+  const auto nWrite = fwrite(
+    output_buffer.data(), sizeof(uint8_t),
+    output_buffer.size(), file_pointer);
+
+  if (nWrite != output_buffer.size()) {
+    ROSBAG2_COMPRESSION_LOG_ERROR_STREAM("Bytes written (" <<
+      nWrite << " != output_buffer_length (" << output_buffer.size() <<
+      ")!");
+    // An error indicator is set by fwrite, so the following check will throw.
+  }
+
+  if (ferror(file_pointer)) {
+    fclose(file_pointer);
+
+    std::stringstream errmsg;
+    errmsg << "Unable to write decompressed data to file: \"" << uri << "\"!";
+
+    throw std::runtime_error{errmsg.str()};
+  }
+
+  fclose(file_pointer);
+}
+
+/**
+ * Checks compression_result and throws a runtime_error if there was a ZSTD error.
+ */
+void throw_on_zstd_error(const size_t compression_result)
+{
+  if (ZSTD_isError(compression_result)) {
+    std::stringstream error;
+    error << "ZSTD decompression error: " << ZSTD_getErrorName(compression_result);
+
+    throw std::runtime_error{error.str()};
+  }
+}
+
+/**
+ * Checks if the frame content is valid.
+ */
+void check_frame_content(const size_t frame_content)
```

Comment:
given that we have `throw_on_zstd_error` I think this function could be renamed accordingly as it really does only throwing on error.